mysql-test/suite/rpl/r/rpl_show_slave_hosts.result

```diff
@@ -129,13 +129,114 @@ Server_id	Host	Port	Master_id	Gtid_State_Sent	Gtid_State_Ack
+Variable_name	Value
+rpl_semi_sync_master_enabled	ON
+SHOW REPLICA HOSTS;
+Server_id	Host	Port	Master_id	Gtid_State_Sent	Gtid_State_Ack
+3	slave2	SLAVE_PORT	1	0-1-4	0-1-4
+2	localhost	SLAVE_PORT	1	0-1-4	0-1-4
+INSERT INTO t VALUES (505);
+show status like 'Rpl_semi_sync_master_yes_tx';
+Variable_name	Value
+Rpl_semi_sync_master_yes_tx	4
+# Q1: Here we need to get the `Gtid_State_Sent=0-1-5 != Gtid_State_Ack=0-1-4` for server_id=2 OR MAYBE THIS IS GOOD,
+# since IO thread of replicas is ON ?
+# Currently we observed the same `Gtid_State_Sent=Gtid_State_Ack=0-1-5`
+SHOW REPLICA HOSTS;
+Server_id	Host	Port	Master_id	Gtid_State_Sent	Gtid_State_Ack
+3	slave2	SLAVE_PORT	1	0-1-5	0-1-5
+2	localhost	SLAVE_PORT	1	0-1-5	0-1-5
+connection slave;
+STOP SLAVE IO_THREAD;
+include/wait_for_slave_io_to_stop.inc
+show status like 'rpl_semi_sync_slave%';
+Variable_name	Value
+Rpl_semi_sync_slave_send_ack	5
+Rpl_semi_sync_slave_status	OFF
+show variables like 'rpl_semi_sync_slave_enabled';
+Variable_name	Value
+rpl_semi_sync_slave_enabled	ON
+connection master;
+show status like "Slaves_connected";
+Variable_name	Value
+Slaves_connected	1
+show status like 'Rpl_semi_sync_master_clients';
+Variable_name	Value
+Rpl_semi_sync_master_clients	1
+show variables like 'rpl_semi_sync_slave_enabled';
+Variable_name	Value
+rpl_semi_sync_slave_enabled	OFF
+# Q2: Here we don't get any row for server_id=2, but we should? we get only `Gtid_State_Sent= Gtid_State_Ack=0-1-5` for server_id=3
+SHOW REPLICA HOSTS;
+Server_id	Host	Port	Master_id	Gtid_State_Sent	Gtid_State_Ack
+3	slave2	SLAVE_PORT	1	0-1-5	0-1-5
+# Write new event when IO thread is stopped
+INSERT INTO t VALUES (505);
+# Writing new events will affect only on replica server_id=3 Gtid_State_Sent= Gtid_State_Ack=0-1-6`
+SHOW REPLICA HOSTS;
+Server_id	Host	Port	Master_id	Gtid_State_Sent	Gtid_State_Ack
+3	slave2	SLAVE_PORT	1	0-1-6	0-1-6
+connection slave;
+START SLAVE IO_THREAD;
+include/wait_for_slave_io_to_start.inc
+# Q*: Stoping and starting the new IO thread will restart slave sent ack count, see below.
+# So we basically want this to be recorded in edge case, right ? 
+# And that is rpl_semi_sync_master_timeout=0 and rpl_semi_sync_master_enabled=TRUE
+# Can we use this as patch instead of injecting the DBUG_SYNC?
+# We are expecting here `Rpl_semi_sync_slave_send_ack == 4` 
+show status like 'rpl_semi_sync_slave%';
+Variable_name	Value
+Rpl_semi_sync_slave_send_ack	0
+Rpl_semi_sync_slave_status	ON
+show variables like 'rpl_semi_sync_slave_enabled';
+Variable_name	Value
+rpl_semi_sync_slave_enabled	ON
+connection master;
+show status like "Slaves_connected";
+Variable_name	Value
+Slaves_connected	2
+show status like 'Rpl_semi_sync_master_clients';
+Variable_name	Value
+Rpl_semi_sync_master_clients	2
+# Q3: here we get  `Gtid_State_Sent= 0-1-5 != Gtid_State_Ack=''` (but Gtid_State_Ack should be 0-1-4 right?)
+# So should we get for server_id=2 `Gtid_State_Sent= 0-1-6 != Gtid_State_Ack=0-1-4` ?
```

Comment:
I think your question ties into the DEBUG_SYNC answer. By keeping the IO thread on, and using DEBUG_SYNC to only pausing it (thereby maintaining the binlog_dump_thread connection), this would be the case when the slave just had not yet replied with the `ACK` (but again, only on the active transaction).